### ISSUE TITLE
Configs: Make sure that the stylistic config doesn't conflict with isolatedDeclarations

### ISSUE THREAD

**Issue by blickly** [typescript-eslint/typescript-eslint#10001]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22preset+config+change%22+) and found none that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Description

The stylistic config includes the https://typescript-eslint.io/rules/consistent-generic-constructors/ rule, which (in it's default configuration) conflicts with the requirements of TS 5.5's new [isolatedDeclarations mode]( https://devblogs.microsoft.com/typescript/announcing-typescript-5-5/#using-isolateddeclarations).

In isolatedDeclarations mode, the following code is illegal:
```
export const s = new Set<string>();
```
giving an error: "Variable must have an explicit type annotation with --isolatedDeclarations."
[TypeScript Playground](https://www.typescriptlang.org/play/?isolatedDeclarations=true#code/FAUwHgDg9gTgLgAgMZQHYGdHoIwILwKogDuCAyiHADyYwCWqA5gHwAUAlANyiSyIoYsAJgBc5SjTj0mzfIRLi4HbsCA)

Whereas in ESLint, the opposite pattern is disallowed by the `consistent-generic-constructors` from the stylistic config:
```
export const s: Set<string> = new Set();
```
gives a different error: "The generic type arguments should be specified as part of the constructor type arguments."
[ESLint Playground](https://typescript-eslint.io/play#ts=5.5.2&fileType=.ts&code=FAUwHgDg9gTgLgAgMZQHYGdHoIwILwKogDuCAyiHADyYwCWqA5gHwAUAlANyiSyIoYsAJgBc5SjTj0mzfIRLi4HbsCA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6MgeyeUuX0Ra0A5gMRUytTt3xwy%2BDtFQZIY6AsjgwAXxBagA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3QAacDUjISsVAEM86ACYARfosxLCpCmFHopAXxDGgA&tokens=false)

Ideally, users using both the stylistic config and the isolatedDeclartions tsconfig option should have a less frustrating out-of-the-box experience.  I think ESLint devs are better positioned to make the call on what that should look like, but I few options that I can think of would be:

1. Turn off `consistent-generic-constructors` by default in the stylistic config, or
2. Change the default behavior of `consistent-generic-constructors` to avoid running on isolatedDeclarations projects, or
3. Change the default behavior to distinguish local and exported declarations so it can give errors that don't contradict the isolatedDeclarations ones.

### Impacted Configurations

stylistic

**Comment by Josh-Cena** [comment]
The stylistic config is, well, _stylistic_. The opinions are what the majority of TS codebases would agree on based on readability, conciseness, and consistency. Now there would always be TS options that make you write code in a different style, and ID often enforces a style that's foreign to non-ID codebases. I don't think we should change the stylistic config to accommodate ID; rather, we could make a config that turns off rules that conflict with ID (for example, in the current world, ID is basically the same as `consistent-generic-constructor: ["error", "type-annotation"]`, so you gain nothing by enabling this ESLint rule).

However, I'm coming over from https://github.com/microsoft/TypeScript/issues/60010, and I think if https://github.com/microsoft/TypeScript/issues/60010 gets accepted, this issue may be completed obsoleted.

**Comment by blickly** [comment]
That makes sense to me.  I think the thrust of my issue is that ID users should have *some* way to turn on "stylistic" linting without needing to make the individual rule decisions themselves.  There are multiple ways to provide that, but having a "stylistic for ID projects" config seems like a perfectly cromulent way to achieve that (as would https://github.com/microsoft/TypeScript/issues/60010, although I'm less clear how feasible that would be to land)

Minor nit: The ID option currently requires type annotations only for exported declarations, so it's a still a style call how to handle local declarations and non-exported module-level declarations.

**Comment by JoshuaKGoldberg** [comment]
As awesome and beneficial and exciting as `isolatedDeclarations` is, a very small percentage of TypeScript projects use it today. It's primarily targeted to large monorepos & such to help with build performance. Their constraints are quite a bit more severe than what most, smaller TypeScript projects go with.

Broadly lessening the usefulness of `consistent-generic-constructors` in the stylistic config to handle those ID repos would be IMO a net negative. It's a legitimately useful _stylistic_ rule that a lot of folks find benefit from.

However, lint rules do have the ability to read compiler options. I think we can consider it a bug that `consistent-generic-constructors` produces a report->fix that causes a TypeScript report when ID is enabled. 

Proposal (cc @typescript-eslint/triage-team): how about we retarget this issue & marking as accepting PRs to make the rule respect `compilerOptions.isolatedDeclarations`?

**Comment by bradzacher** [comment]
I'm personally at a -1

At best I think we could offer a config to layer on top to disable any rules that conflict with it.

But I would probably say "this requirement applies to less than 10% of configs so it's not worth us building a config to make it easier".

**Comment by kirkwaiblinger** [comment]
> Proposal (cc @typescript-eslint/triage-team): how about we retarget this issue & marking as accepting PRs to make the rule respect `compilerOptions.isolatedDeclarations`?

What does this look like concretely? Issue an error with incompatible options, like, eg, prefer-nullish-coalescing without strictNullChecks? Or something else?

**Comment by JoshuaKGoldberg** [comment]
I was thinking that `consistent-generic-constructors` should avoid emitting the report about switching generic type arguments when ID is enabled.

**Comment by kirkwaiblinger** [comment]
@JoshuaKGoldberg 

Should ID users just use the 'type-annotation' option, though, rather than the 'constructor' option? Would the 'constructor' option with ID ever report under this proposal?

**Comment by bradzacher** [comment]
My 2c is that ID users are rare enough that we shouldn't look to support it aside from making it possible for rules to be used with it when it makes sense.
Some rules are going to conflict with it irreconcilably (eg `no-inferrable-types`) and some will complement it (eg `explicit-module-boundary-types`).

We should just make sure that rules that could work with it have the option to do so.

But I don't think we should change the config around it.

**Comment by JoshuaKGoldberg** [comment]
> Should ID users just use the 'type-annotation' option, though, rather than the 'constructor' option? Would the 'constructor' option with ID ever report under this proposal?

ID is only relevant for `export`ed stuff. Anything local is fine.

```ts
export const checked = new Set<string>();

export async function main() {
  const ignored = new Set<string>();
}
```

> We should just make sure that rules that could work with it have the option to do so.

Sounds like we're generally 👍 on this - so from my interpretation, the only next step would be to make `consistent-generic-constructors` and `no-inferrable-types` ignore top-level `export`ed stuff when ID is on. Right?

**Comment by kirkwaiblinger** [comment]
> > Should ID users just use the 'type-annotation' option, though, rather than the 'constructor' option? Would the 'constructor' option with ID ever report under this proposal?
> 
> ID is only relevant for `export`ed stuff. Anything local is fine.

Ah! gotcha, that was the missing part for me. Thanks!

> > We should just make sure that rules that could work with it have the option to do so.
> 
> Sounds like we're generally 👍 on this - so from my interpretation, the only next step would be to make `consistent-generic-constructors` and `no-inferrable-types` ignore top-level `export`ed stuff when ID is on. Right?

👍 from me

**Comment by bradzacher** [comment]
> the only next step would be to make consistent-generic-constructors and no-inferrable-types ignore top-level exported stuff when ID is on. Right?

To determine if ID is on you need type-aware linting so we can get the tsconfig. But both of these rules are not type-aware.

Which would mean we need to add a parser option which the user can manually set or we prefill if type-aware linting is on (just like the emitDecoratorMetadata setup).

Seems like a lot of effort for what is right now a pretty niche usecase, TBH.

**Comment by JoshuaKGoldberg** [comment]
Is it more complex than https://github.com/typescript-eslint/typescript-eslint/pull/10499?

I think it's fair to assume ID is going to slowly get more popular over time. Especially given tools like [oxbuild](https://github.com/DonIsaac/oxbuild) and [unplugin-isolated-decl](https://github.com/unplugin/unplugin-isolated-decl) getting more popular that generate `.d.ts` files without `tsc`. [jsr also requires "fast" types](https://jsr.io/docs/about-slow-types), which is explicitly pretty close to `isolatedDeclarations`. 

tl;dr: I'm pretty solidly +1 on taking this cost.

**Comment by JoshuaKGoldberg** [comment]
It looks like we have consensus on not changing the stylistic config itself to support isolatedDeclarations (ID). Multiple of us are leaning towards instead having individual rules understand ID and not produce conflicting reports. So I'm going to go ahead and close this issue out in favor of splitting out:

* #10571
* #10572

Thanks all!
